File: src/ios/AdMobNextGen.m
#import "AdMobNextGen.h"
#import "ConsentExecutor.h"
#import "GlobalSettingsExecutor.h"
#import "BannerExecutor.h"
#import "InterstitialExecutor.h"
#import "RewardedExecutor.h"
#import "AppOpenAdExecutor.h"
#import "RewardedInterstitialExecutor.h"

@interface AdMobNextGen()
@property (nonatomic, strong) ConsentExecutor *consentExecutor;
@property (nonatomic, strong) GlobalSettingsExecutor *globalSettingsExecutor;
@property (nonatomic, strong) BannerExecutor *bannerExecutor;
@property (nonatomic, strong) InterstitialExecutor *interstitialExecutor; 
@property (nonatomic, strong) RewardedExecutor *rewardedExecutor;
@property (nonatomic, strong) RewardedInterstitialExecutor *rewardedInterstitialExecutor;
@end

@implementation AdMobNextGen

- (void)pluginInitialize {
    [super pluginInitialize];
    self.consentExecutor = [[ConsentExecutor alloc] initWithPlugin:self];
    self.globalSettingsExecutor = [[GlobalSettingsExecutor alloc] initWithPlugin:self];
    self.bannerExecutor = [[BannerExecutor alloc] initWithPlugin:self];
    self.interstitialExecutor = [[InterstitialExecutor alloc] initWithPlugin:self]; 
    self.rewardedExecutor = [[RewardedExecutor alloc] initWithPlugin:self];
    [[AppOpenAdExecutor sharedInstance] initializeWithPlugin:self];
    self.rewardedInterstitialExecutor = [[RewardedInterstitialExecutor alloc] initWithPlugin:self];
}

- (void)initialize:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        [[GADMobileAds sharedInstance] startWithCompletionHandler:^(GADInitializationStatus * _Nonnull status) {
            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"Initialization complete."];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }];
    }];
}

#pragma mark - Consent (UMP) Routing

- (void)requestConsentInfo:(CDVInvokedUrlCommand*)command {
    NSDictionary *options = nil;
    if (command.arguments.count > 0) {
        options = [command.arguments objectAtIndex:0];
    }
    [self.consentExecutor requestConsentInfo:options command:command];
}

- (void)showPrivacyOptionsForm:(CDVInvokedUrlCommand*)command {
    [self.consentExecutor showPrivacyOptionsForm:command];
}

- (void)getTCData:(CDVInvokedUrlCommand*)command {
    [self.consentExecutor getTCData:command];
}

- (void)canRequestAds:(CDVInvokedUrlCommand *)command {
    [self.consentExecutor canRequestAds:command];
}

- (void)requestTrackingAuthorization:(CDVInvokedUrlCommand *)command {
    [self.consentExecutor requestTrackingAuthorization:command];
}

- (void)getTrackingAuthorizationStatus:(CDVInvokedUrlCommand *)command {
    [self.consentExecutor getTrackingAuthorizationStatus:command];
}

#pragma mark - Global Settings Routing

- (void)setAppVolume:(CDVInvokedUrlCommand*)command {
    [self.globalSettingsExecutor setAppVolume:command];
}

- (void)setAppMuted:(CDVInvokedUrlCommand*)command {
    [self.globalSettingsExecutor setAppMuted:command];
}

- (void)setRequestConfiguration:(CDVInvokedUrlCommand*)command {
    [self.globalSettingsExecutor setRequestConfiguration:command];
}

#pragma mark - Banner Routing

- (void)createBanner:(CDVInvokedUrlCommand*)command {
    NSDictionary *options = [command.arguments objectAtIndex:0];
    if (options != nil && [options isKindOfClass:[NSDictionary class]]) {
        [self.bannerExecutor createBanner:options command:command];
    } else {
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Invalid options object."];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }
}

- (void)showBanner:(CDVInvokedUrlCommand*)command {
    [self.bannerExecutor showBanner:command];
}

- (void)hideBanner:(CDVInvokedUrlCommand*)command {
    [self.bannerExecutor hideBanner:command];
}

- (void)removeBanner:(CDVInvokedUrlCommand*)command {
    [self.bannerExecutor removeBanner:command];
}

#pragma mark - Interstitial Routing

- (void)createInterstitial:(CDVInvokedUrlCommand*)command {
    NSDictionary *options = [command.arguments objectAtIndex:0];
    if (options != nil && [options isKindOfClass:[NSDictionary class]]) {
        [self.interstitialExecutor createInterstitial:options command:command];
    } else {
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Invalid options object."];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }
}

- (void)showInterstitial:(CDVInvokedUrlCommand*)command {
    [self.interstitialExecutor showInterstitial:command];
}

- (void)createRewarded:(CDVInvokedUrlCommand*)command {
    NSDictionary *options = [command.arguments objectAtIndex:0];
    if (options != nil) {
        [self.rewardedExecutor createRewarded:options command:command];
    }
}

- (void)showRewarded:(CDVInvokedUrlCommand*)command {
    [self.rewardedExecutor showRewarded:command];
}

- (void)loadAppOpenAd:(CDVInvokedUrlCommand*)command {
    NSDictionary *options = [command.arguments objectAtIndex:0];
    if (options != nil) {
        [[AppOpenAdExecutor sharedInstance] loadAppOpenAd:options command:command];
    }
}

- (void)showAppOpenAd:(CDVInvokedUrlCommand*)command {
    [[AppOpenAdExecutor sharedInstance] showAppOpenAd:command];
}

- (void)createRewardedInterstitial:(CDVInvokedUrlCommand*)command {
    NSDictionary *options = [command.arguments objectAtIndex:0];
    if (options != nil) {
        [self.rewardedInterstitialExecutor createRewardedInterstitial:options command:command];
    }
}

- (void)showRewardedInterstitial:(CDVInvokedUrlCommand*)command {
    [self.rewardedInterstitialExecutor showRewardedInterstitial:command];
}

#pragma mark - Event Emitter

- (void)fireEvent:(NSString *)obj event:(NSString *)eventName withData:(NSString *)jsonStr {
    NSString* js;
    if(obj && [obj isEqualToString:@"window"]) {
        js = [NSString stringWithFormat:@"var evt=document.createEvent(\"UIEvents\");evt.initUIEvent(\"%@\",true,false,window,0);window.dispatchEvent(evt);", eventName];
    } else if(jsonStr && [jsonStr length] > 0) {
        js = [NSString stringWithFormat:@"javascript:cordova.fireDocumentEvent('%@',%@);", eventName, jsonStr];
    } else {
        js = [NSString stringWithFormat:@"javascript:cordova.fireDocumentEvent('%@');", eventName];
    }

    dispatch_async(dispatch_get_main_queue(), ^{
        [self.commandDelegate evalJs:js];
    });
}

@end

File: www/admob-nextgen.js
var exec = require('cordova/exec');

var AdMobNextGen = {

    /*
     Created by EMI INDO So on Frb 10, 2026
    */

    setAppVolume: function (volume, success, error) {
        exec(success, error, 'AdMobNextGen', 'setAppVolume', [volume]);
    },

    setAppMuted: function (muted, success, error) {
        exec(success, error, 'AdMobNextGen', 'setAppMuted', [muted]);
    },

    setRequestConfiguration: function (config, success, error) {
        exec(success, error, 'AdMobNextGen', 'setRequestConfiguration', [config]);
    },

    requestConsentInfo: function (options, success, error) {
        exec(success, error, 'AdMobNextGen', 'requestConsentInfo', [options]);
    },

    showPrivacyOptionsForm: function (success, error) {
        exec(success, error, 'AdMobNextGen', 'showPrivacyOptionsForm', []);
    },

    getTCData: function (success, error) {
        exec(success, error, 'AdMobNextGen', 'getTCData', []);
    },

    requestTrackingAuthorization: function (success, error) {
        exec(success, error, 'AdMobNextGen', 'requestTrackingAuthorization', []);
    },

    getTrackingAuthorizationStatus: function (success, error) {
        exec(success, error, 'AdMobNextGen', 'getTrackingAuthorizationStatus', []);
    },

    // Inisialisasi SDK
    initialize: function (config, success, error) {
        var args = [];
        var successCb = success;
        var errorCb = error;

        if (typeof config === 'function') {
            successCb = config;
            errorCb = success;
        } else if (config && typeof config === 'object') {
            args = [config];
        }

        exec(successCb, errorCb, 'AdMobNextGen', 'initialize', args);
    },

    createBanner: function (options, success, error) {
        var isCapacitorEnvironment = typeof window.Capacitor !== 'undefined';
        
        options = options || {};
        options.isCapacitor = isCapacitorEnvironment;

        exec(success, error, 'AdMobNextGen', 'createBanner', [options]);
    },

    hideBanner: function (success, error) {
        exec(success, error, 'AdMobNextGen', 'hideBanner', []);
    },

    showBanner: function (success, error) {
        exec(success, error, 'AdMobNextGen', 'showBanner', []);
    },

    removeBanner: function (success, error) {
        exec(success, error, 'AdMobNextGen', 'removeBanner', []);
    },

    createInterstitial: function (options, successEvent, error) {
        exec(successEvent, error, 'AdMobNextGen', 'createInterstitial', [options]);
    },

    showInterstitial: function (success, error) {
        exec(success, error, 'AdMobNextGen', 'showInterstitial', []);
    },

    createRewarded: function (options, successEvent, error) {
        exec(successEvent, error, 'AdMobNextGen', 'createRewarded', [options]);
    },

    showRewarded: function (success, error) {
        exec(success, error, 'AdMobNextGen', 'showRewarded', []);
    },

    createRewardedInterstitial: function (options, success, error) {
        exec(success, error, 'AdMobNextGen', 'createRewardedInterstitial', [options]);
    },

    showRewardedInterstitial: function (success, error) {
        exec(success, error, 'AdMobNextGen', 'showRewardedInterstitial', []);
    },

    loadAppOpenAd: function (options, successEvent, error) {
        exec(successEvent, error, 'AdMobNextGen', 'loadAppOpenAd', [options]);
    },

    showAppOpenAd: function (success, error) {
        exec(success, error, 'AdMobNextGen', 'showAppOpenAd', []);
    },

    createNativeAd: function (options, successEvent, error) {
        exec(successEvent, error, 'AdMobNextGen', 'createNativeAd', [options]);
    },

    removeNativeAd: function (success, error) {
        exec(success, error, 'AdMobNextGen', 'removeNativeAd', []);
    },

    startBannerPreload: function (options, success, error) {
        var isCapacitorEnvironment = typeof window.Capacitor !== 'undefined';
        
        options = options || {};
        options.isCapacitor = isCapacitorEnvironment;

        exec(success, error, 'AdMobNextGen', 'startBannerPreload', [options]);
    },

    showPreloadedBanner: function (success, error) {
        exec(success, error, 'AdMobNextGen', 'showPreloadedBanner', []);
    },

    stopBannerPreload: function (success, error) {
        exec(success, error, 'AdMobNextGen', 'stopBannerPreload', []);
    },

    startAppOpenPreload: function (options, success, error) {
        exec(success, error, 'AdMobNextGen', 'startAppOpenPreload', [options]);
    },

    showPreloadedAppOpenAd: function (success, error) {
        exec(success, error, 'AdMobNextGen', 'showPreloadedAppOpenAd', []);
    },

    isAppOpenAdAvailable: function (success, error) {
        exec(success, error, 'AdMobNextGen', 'isAppOpenAdAvailable', []);
    }
};

module.exports = AdMobNextGen;